### PR TITLE
KAFKA-15129;[1/N] Remove metrics in LogCleanerManager when LogCleaner shutdown

### PR DESCRIPTION
**Pull request by hudeqi**
This pr is used to remove the metrics in LogCleanerManager when logCleaner is closed.
This pr has passed the corresponding unit test, and it is part of KAFKA-15129.

### CODE REVIEW

**Comment by hudeqi** [comment]
I have submitted some PRs (two others: https://github.com/apache/kafka/pull/13926、https://github.com/apache/kafka/pull/13929) on the topic of "clean metric", please help to review, thank you! @clolov @divijvaidya

**Review by clolov — APPROVED** [review]
Beyond wanting the variable names to start with lower-case unless there is a good reason to not do so, I am happy with this!

File: core/src/main/scala/kafka/log/LogCleanerManager.scala
Comment (clolov) [inline core/src/main/scala/kafka/log/LogCleanerManager.scala:573]: Is there a reason why decided to have these variable names start with a capital letter?

Comment (hudeqi) [reply core/src/main/scala/kafka/log/LogCleanerManager.scala:573]: > Is there a reason why decided to have these variable names start with a capital letter?

Learned scala writing rules from discussions like PR： https://github.com/apache/kafka/pull/13623#discussion_r1182593025  @clolov

File: core/src/main/scala/kafka/log/LogCleanerManager.scala
Comment (divijvaidya) [inline core/src/main/scala/kafka/log/LogCleanerManager.scala]: could we move the `Map("logDirectory" -> dir.getAbsolutePath` part to a val metricTag and then re-use it at both locations.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: does this work? In our case we will call this multiple times with same metric Name. The any() will match all such occurrences whereas we are saying that this invocation occurs only 1 time (default is times(1), when not specified)

Please help me understand why it's working.

Comment (clolov) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: Sorry, why do you think we will call this multiple times with the same metric name? The way I read this code is that we will get the keys from GaugeMetricNameWithTag and GaugeMetricNameNoTag (which are unique because of the nature of a set and map) and for each one we will verify that it is called only once.

Comment (divijvaidya) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: GaugeMetricNameWithTag can contain multiple metrics with same name but different tags. Let's say `UncleanablePartitionsCountMetricName ` is the name and tag will be directory path. Over here, we are verifying that there is an exactly 1 call to newGauge() with `UncleanablePartitionsCountMetricName` as the metricname. But that is not correct because we are actually calling `newGauge` for this key name numDirs times. 

Hence, this test should ideally fail. What am I missing here?

Comment (hudeqi) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: > GaugeMetricNameWithTag can contain multiple metrics with same name but different tags. Let's say `UncleanablePartitionsCountMetricName ` is the name and tag will be directory path. Over here, we are verifying that there is an exactly 1 call to newGauge() with `UncleanablePartitionsCountMetricName` as the metricname. But that is not correct because we are actually calling `newGauge` for this key name numDirs times.
> 
> Hence, this test should ideally fail. What am I missing here?

I think the reason is: "logDirs = Array(TestUtils.tempDir())", means that the `numDirs` size is 1. @divijvaidya

Comment (divijvaidya) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: ah ok. Can we please unit test this will multi dirs since the metric addition/removal here is dependent on number of directories and we want to ensure it works correctly for multiple dir case as well.

File: core/src/main/scala/kafka/log/LogCleanerManager.scala
Comment (hudeqi) [reply core/src/main/scala/kafka/log/LogCleanerManager.scala]: committed the update.

File: core/src/test/scala/unit/kafka/log/LogCleanerTest.scala
Comment (hudeqi) [reply core/src/test/scala/unit/kafka/log/LogCleanerTest.scala]: I have updated this. @divijvaidya

**Review by divijvaidya — APPROVED** [review]
Looks good! I will wait for CI to complete before we merge this.

**Comment by hudeqi** [comment]
Hi, can we merge this and review the next three PR of the topic? @divijvaidya

**Comment by divijvaidya** [comment]
Unrelated test failures, merging this in. @hudeqi I have rest of the PRs in my queue and will reach to them in a few days.
```
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.connect.mirror.integration/IdentityReplicationIntegrationTest/Build___JDK_20_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_20_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_20_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow___2/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_17_and_Scala_2_13___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.connect.mirror.integration/MirrorConnectorsIntegrationExactlyOnceTest/Build___JDK_8_and_Scala_2_12___testOffsetTranslationBehindReplicationFlow__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders__/)
[Build / JDK 8 and Scala 2.12 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_8_and_Scala_2_12___testBalancePartitionLeaders___2/)
[Build / JDK 11 and Scala 2.13 / kafka.admin.TopicCommandIntegrationTest.testDescribeUnderMinIsrPartitionsMixed(String).quorum=zk](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/kafka.admin/TopicCommandIntegrationTest/Build___JDK_11_and_Scala_2_13___testDescribeUnderMinIsrPartitionsMixed_String__quorum_zk/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testBalancePartitionLeaders__/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-13924/5/testReport/junit/org.apache.kafka.trogdor.coordinator/CoordinatorTest/Build___JDK_11_and_Scala_2_13___testTaskRequestWithOldStartMsGetsUpdated__/)
```

File: core/src/main/scala/kafka/log/LogCleanerManager.scala
Comment (kamalcph) [inline core/src/main/scala/kafka/log/LogCleanerManager.scala:578]: Why variable names are not camelCase? 
 
@hudeqi @divijvaidya

Comment (divijvaidya) [reply core/src/main/scala/kafka/log/LogCleanerManager.scala:578]: From https://kafka.apache.org/coding-guide:

`Constants should be camel case with an initial capital LikeThis not LIKE_THIS.`

Comment (divijvaidya) [reply core/src/main/scala/kafka/log/LogCleanerManager.scala:578]: Let me take that back.

If it were a def function, we would probably have used camel case. For constants such as `TimeSinceLastRunMsMetricName` definitely first capital letter. But for variable names such as `GaugeMetricNameWithTag` we should probably use small case. 

Also, this `GaugeMetricNameWithTag` is an instance specific member and does not belong in companion object. @hudeqi would you please move this into the class itself, since it stores a state which is specific to the instance.

Comment (hudeqi) [reply core/src/main/scala/kafka/log/LogCleanerManager.scala:578]: Thanks! @divijvaidya @kamalcph. I will pay attention to this issue in subsequent PRs of this type. I have submitted the update of this PR to this minor PR(https://github.com/apache/kafka/pull/13958), thanks for the review.

File: core/src/main/scala/kafka/log/LogCleanerManager.scala
Comment (dajac) [inline core/src/main/scala/kafka/log/LogCleanerManager.scala:100]: Hum... This does not seem safe. I just noticed the following failure in CI:

```
Error
org.apache.kafka.server.fault.FaultHandlerException: fatalFaultHandler: Error starting LogManager: null
Stacktrace
org.apache.kafka.server.fault.FaultHandlerException: fatalFaultHandler: Error starting LogManager: null
	at java.base@17.0.7/java.util.HashMap.computeIfAbsent(HashMap.java:1221)
	at app//kafka.log.LogCleanerManager.$anonfun$new$1(LogCleanerManager.scala:100)
	at app//kafka.log.LogCleanerManager.$anonfun$new$1$adapted(LogCleanerManager.scala:94)
	at app//scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at app//scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at app//scala.collection.AbstractIterable.foreach(Iterable.scala:933)
	at app//kafka.log.LogCleanerManager.<init>(LogCleanerManager.scala:94)
	at app//kafka.log.LogCleaner.<init>(LogCleaner.scala:109)
	at app//kafka.log.LogManager.startupWithConfigOverrides(LogManager.scala:571)
	at app//kafka.log.LogManager.startup(LogManager.scala:504)
	at app//kafka.server.metadata.BrokerMetadataPublisher.initializeManagers(BrokerMetadataPublisher.scala:350)
	at app//kafka.server.metadata.BrokerMetadataPublisher.onMetadataUpdate(BrokerMetadataPublisher.scala:162)
	at app//org.apache.kafka.image.loader.MetadataLoader.initializeNewPublishers(MetadataLoader.java:301)
	at app//org.apache.kafka.image.loader.MetadataLoader.lambda$scheduleInitializeNewPublishers$0(MetadataLoader.java:258)
	at app//org.apache.kafka.queue.KafkaEventQueue$EventContext.run(KafkaEventQueue.java:127)
	at app//org.apache.kafka.queue.KafkaEventQueue$EventHandler.handleEvents(KafkaEventQueue.java:210)
	at app//org.apache.kafka.queue.KafkaEventQueue$EventHandler.run(KafkaEventQueue.java:181)
	at java.base@17.0.7/java.lang.Thread.run(Thread.java:833)
Caused by: java.util.ConcurrentModificationException
	... 18 more
```

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13954/6/tests

cc @divijvaidya

Comment (dajac) [reply core/src/main/scala/kafka/log/LogCleanerManager.scala:100]: We may want to revert this commit while we fix this.

Comment (divijvaidya) [reply core/src/main/scala/kafka/log/LogCleanerManager.scala:100]: We recognized this error and should be fixed by https://github.com/apache/kafka/pull/13958 

The problem is that we accidentally made this map as a static (in companion object) whereas it should have been a instance level member. Due to it's static nature, it is getting concurrent modification exception.

**Pull request merged**
